Add input plugin for myepisodes.com show list

oxc opened the pull request:
This pull request contains the file I've already attached to [trac-ticket #243](http://flexget.com/ticket/243:comment9):

myepisodes_list.py​ contains an input plugin that lists all ever-watched shows from myepisodes.com (all that show up on the "Time I've Wasted" page). The plugin has been based on code from the myepisodes service plugin and other input plugins, so I hope I got the coding style right.

However, I did not abstract any myepisodes api that can be shared between the two plugins, something that should probably be done (but rather by someone with more flexget experience than me).

stevezau commented on flexget/plugins/input/myepisodes_list.py:
```diff
@@ -0,0 +1,98 @@
+        wasted_url = "%s/%s" % (URL, "life_wasted.php")
+        log.debug('Fetching "Time I\'ve wasted" page: %s' % wasted_url)
+
+        page = session.get(wasted_url).content
```
any reason why task.requests is not used here?

stevezau commented on flexget/plugins/input/myepisodes_list.py:
```diff
@@ -0,0 +1,98 @@
+        log.debug('Fetching "Time I\'ve wasted" page: %s' % wasted_url)
+
+        page = session.get(wasted_url).content
+        soup = get_soup(page)
```
Do we need some exception handling here?

stevezau commented:
@oxc i've made a few comments, other then that the code looks ok.

oxc commented:
@stevezau thanks for the review. I've updated the change to include your comments. The reason I did not use `task.requests` was that I didn't know about it ;) I've added a rather defensive error handling, let me know if this is too much.

Also, I've changed the source of the list to a different page on myepisodes.com, which is much more suitable for this task. It includes new shows that have not been watched yet at all. Additionally, it allows me to omit the ignored shows, which is the new default; they can be added back by setting the new `include_ignored` to True.

Please review, thanks.

oxc commented:
Forgot to include the new option in the docstring, this is now fixed.

oxc commented:
Cleaned up some imports and simplified the shows URL creation

stevezau commented on flexget/plugins/input/myepisodes_list.py:
```diff
@@ -0,0 +1,99 @@
+                'password': password,
+                'action': 'Login'
+            }
+            loginsrc = task.requests.post(URL + 'login.php', data=params).content
```
Need some exception handling here... What if we get a http 5xx, 3xx or timeout error?

stevezau replied:
FYI http://docs.python-requests.org/en/latest/user/quickstart/#errors-and-exceptions

gazpachoking replied:
The internet decorator will actually handle it as is, though the error message might not be as nice as explicit error handling.

stevezau commented on flexget/plugins/input/myepisodes_list.py:
```diff
@@ -0,0 +1,99 @@
+                                 'your account data or see if the site is down.'), log)
+
+
+        page = task.requests.get(URL + "shows.php?type=manage").content
```
exception handling needed for http errors and timeouts

stevezau commented on flexget/plugins/input/myepisodes_list.py:
```diff
@@ -0,0 +1,99 @@
+                else:
+                    log.debug('Invalid entry created? %s' % entry)
+
+        except Exception as e:
```
We shouldn't use blanket exception handling. check here for more info on soup excpetions

http://www.crummy.com/software/BeautifulSoup/bs4/doc/#errors-when-parsing-a-document

oxc commented:
Isn't the @plugin.internet decorator supposed to handle the request exceptions you mentioned?

There's really nothing the plugin can do if any of the requests or parsing the site fails. Returning an empty list in that case does not sound right either.

stevezau commented:
Yes you are right, sorry didn't see @plugin.internet decorator.

Probably should change that blanket exception though?

oxc commented:
Yes, I will do that.

stevezau commented:
Also, what are the advantages of using www.myepisodes.com over trakt or something similar? The www.myepisodes.com website looks outdated, the last accouncement was almost 10 years ago.

oxc commented:
The advantage is that it has my show list in it :-)

I even tried migrating it to Trakt just for Flexget, but it has a different workflow and the interface, albeit beautiful, is much more unprecise and cluttered when it comes to simply having a number of shows of which you want to track acquiring and watching each episode.

I will try again if the site goes down, but until then I'm perfectly happy with myepisdoes.com

stevezau commented:
@oxc i'm really not sure/convinced this should added to the flexget base as i think you might be the only user using it which suggested it might be better as a user plugin?

FYI @gazpachoking is working on a way to create and track movie/series lists internally to flexget.. I'm working on a webui and RESTful api which will eventually allow you to manage your own lists via a mobily friendly ui.. Might be a better solution for you?..

bfd commented:
Well, there is a ticket for it ( http://flexget.com/ticket/243 ), and there already is a myepisodes scrobbler plugin. You should throw out that as well then.
Me for my part, I'm happy to maintain this for my own. I just thought I'd share...

oxc commented:
Well, there is a ticket for it ( http://flexget.com/ticket/243 ), and there already is a myepisodes scrobbler plugin. You should throw out that as well then.
Me for my part, I'm happy to maintain this for my own. I just thought I'd share...

stevezau commented:
@oxc that's all the convincing I need.. Fix that blanket exception and ill merge.

oxc commented:
I've narrowed the scope of the try-block, but I have not changed the exception type. It seems very unreliable what kind of exceptions beautifulsoup actually throws, depending on the parser used. I could not found a single mention of HTMLParser or HTMLParseError in the whole flexget project.

Otherwise, I've only added a warning when the collected entries are empty, since this could mean that the user has not added any shows on myepisodes.com and might be confused by the default flexget message for empty output.

gazpachoking commented:
Thanks! Want to add it to the wiki?

oxc commented:
Done: http://flexget.com/wiki/Plugins/myepisodes_list
